### DISCUSSION TITLE
Trying to create a custom column with a popup containing advanced select field

### DISCUSSION THREAD

**Discussion by nsrosenqvist** [filamentphp/filament#11985] [Help]
### Package

Table builder

### Package Version

v3.2.42

### How can we help you?

I'm trying to create a custom column where the user can click the column in order to open a popover containing a select field with search functionality. I've tried doing it in a variety of ways. I'd like to avoid writing the entire implementation myself and instead reuse existing filament functionality where possible, since supporting everything that the original filament form select input does would lead to something really hard to maintain in the long run. The functionality I want is that when the user clicks the column, then a popover opens with the ability to search for the record they want to associate with the field.

<img width="291" alt="Screenshot 2024-03-22 at 13 37 52" src="https://github.com/filamentphp/filament/assets/1303475/c44dd4e4-73a5-491a-a63b-fb64d54edfda">

The implementation is basically just a custom field, based on the examples I've found for editable columns in the filament tables package, containing a livewire component that makes use of a filament form. I've tried both setting the data as part of saving the field and then dispatching an event to the column class that forces a refresh of the component, as well as only updating the state in the livewire component, not saving the data on the model, but instead save that in the column by passing the data in the event instead. The former is probably a bad idea since I realized that the livewire component is actually the entire table, and not just the column, so refreshing it is probably overkill (did I understand this correctly?). The second way actually appears to work quite well at first glance.

However, when having saved the data in the event handler on the column, and there's any type of column sorting active on the table, then livewire appears to lose track of the components' state (for both methods described above).

<img width="551" alt="Screenshot 2024-03-22 at 13 43 55" src="https://github.com/filamentphp/filament/assets/1303475/2e9d9d7c-97ea-41c5-a840-98414a230457">

I've tried to add `wire:key` to the component, but there is really no iterating that I myself perform, so I fail to see where I would need to add it to make it work correctly. I then bit the bullet and started reimplementing the functionality of the Select form component myself, however, when I realized that the livewire component is actually the table, and not the column, then I don't see how I'd be able to provide the endpoints required for the search functionality used by the select. So I'm back to the original solution again.

This is my current column view:
```blade.php
@php
    extract($getRelationshipArguments());

    $avatar = $getAvatar();
    $avatarAlt = $getAvatarAlt() ?? __("User's avatar");
    $avatarPlaceholder = $getAvatarPlaceholder() ?? '∅';
    $selectPlaceholder = $getSelectPlaceholder() ?? __('No one selected');
    $selectablePlaceholder = $canSelectPlaceholder();
    $searchPrompt = $getSearchPrompt();
    $isDisabled = $isDisabled();

    $tableId = $this->getId();
    $columnName = $column->getName();
    $key = "{$tableId}.table.record.{$recordKey}.column.{$columnName}";
    $formKey = "{$key}.form";

    $state = $getState();
    if ($state instanceof \BackedEnum) {
        $state = $state->value;
    }
    $state = strval($state);
@endphp

<div
    x-data="{
        error: undefined,
        isLoading: false,
        name: @js($getName()),
        recordKey: @js($recordKey),
        state: @js($state),
    }"
    x-init="
        () => {
            Livewire.hook('commit', ({ component, commit, succeed, fail, respond }) => {
                succeed(({ snapshot, effect }) => {
                    $nextTick(() => {
                        if (component.id !== @js($this->getId())) {
                            return
                        }

                        if (! $refs.newState) {
                            return
                        }

                        let newState = $refs.newState.value

                        if (state === newState) {
                            return
                        }

                        state = newState
                    })
                })
            })
        }
    "
    x-on:popup-select="
        if ($event.detail.id !== @js($formKey)) {
            return
        }

        isLoading = true

        const response = await $wire.updateTableColumnState(
            name,
            recordKey,
            $event.detail.state,
        )

        error = response?.error ?? undefined

        if (! error) {
            state = response
        }

        isLoading = false
    "
    {{
        $attributes
            ->merge($getExtraAttributes(), escape: false)
            ->class([
                'mi-user-input', 'mi-avatar-input',
                'w-full', 'h-full',
            ])
    }}
>
    <input
        type="hidden"
        value="{{ str($state)->replace('"', '\\"') }}"
        x-ref="newState"
    >

    <x-filament::dropdown x-ref="panel">
        <x-slot name="trigger" @class(['flex', 'w-full', 'h-full', 'items-center', 'justify-center'])>
            @if($isAvatar())
                @if($avatar)
                    <x-filament::avatar
                        src="{{ $avatar }}"
                        alt="{{ $avatarAlt }}"
                        size="md"
                        circular="true"
                        @class(['mi-action'])
                    />
                @else
                    <x-filament-tables::columns.placeholder>
                        <div @class(['mi-action', 'fi-circular', 'rounded-full', 'leading-7', 'h-8', 'w-8', 'text-center'])>
                            {{ $avatarPlaceholder }}
                        </div>
                    </x-filament-tables::columns.placeholder>
                @endif
            @endif
        </x-slot>

        <div @class(['p-2'])>
            <x-filament::input.wrapper
                :disabled="$isDisabled"
                :alpine-disabled="'isLoading || ' . \Illuminate\Support\Js::from($isDisabled)"
                alpine-valid="error === undefined"
                x-tooltip="
                    error === undefined
                        ? false
                        : {
                            content: error,
                            theme: $store.theme,
                        }
                "
            >
                <livewire:avatar-select
                    :alt="$avatarAlt"
                    :attribute="$getAttribute() ?? $getName()"
                    :name="$getName()"
                    :$recordKey
                    :$avatar
                    :model="$getRecord()"
                    :$avatarPlaceholder
                    :disabled="$isDisabled"
                    :$searchPrompt
                    :$selectPlaceholder
                    :$selectablePlaceholder
                    :$relationshipAvatarAttribute
                    :$relationshipName
                    :$relationshipTitleAttribute
                    :$ignoreRecord
                    :key="$formKey"
                    :formKey="$formKey"
                >
            </x-filament::input.wrapper>
        </div>
    </x-filament::dropdown>
</div>
```

Any ideas? I'm still quite new to filament and livewire in general, so please excuse me if I'm missing something basic. Is there any way one can add column endpoints without completely butchering the `updateState` method for example? This use case can't be unique, can it?

**Comment by shivaRamdeen** [comment]
Did you ever get a solution for this?

**Reply by nsrosenqvist** [reply]
Sorry, no, never got it to work properly and ran out of time to tinker with it :(
